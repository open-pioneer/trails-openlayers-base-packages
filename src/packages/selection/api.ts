// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { Geometry } from "ol/geom";
import type { Projection } from "ol/proj";
import type { Extent } from "ol/extent";
import type { EventSource, Resource } from "@open-pioneer/core";
import { BaseFeature } from "@open-pioneer/map";
import { DeclaredService } from "@open-pioneer/runtime";
import VectorLayer from "ol/layer/Vector";
import Feature from "ol/Feature";

/**
 * The status of a selection source.
 *
 * This is used to indicate whether the source is ready for selection.
 */
export type SelectionSourceStatus = "available" | "unavailable" | SelectionSourceStatusObject;

export type SelectionSourceStatusObject =
    | { kind: "available" }
    | {
          kind: "unavailable";

          /**
           * If the status of this source is unavailable, the reason for this can be stored here.
           *
           * This will be displayed by the user interface.
           *
           * If it is not defined, a default message will be displayed instead.
           */
          reason?: string;
      };

/**
 * Represents a result returned by a spatial selection.
 */
export interface SelectionResult extends BaseFeature {
    /**
     * Geometry of the selection result.
     * One should also specify the {@link projection}.
     */
    geometry: Geometry;
}

/** Options passed to a {@link SelectionSource} when triggering a select. */
export interface SelectionOptions {
    /**
     * The maximum number of selection results to request.
     * The selection component currently only supports a certain amount of results (indicated by this value).
     * If a source results more than `maxResults` results, additional results will be ignored.
     */
    maxResults: number;

    /**
     * The current projection of the map.
     * Useful to return the selection result's geometry in the suitable projection, should they differ.
     */
    mapProjection: Projection;

    /**
     * The signal can be used to detect cancellation.
     *
     * You can pass this signal to builtin functions like `fetch` that automatically
     * support cancellation.
     */
    signal: AbortSignal;
}

/** Events emitted by the {@link SelectionSource}. */
export interface SelectionSourceEvents {
    "changed:status": void;
}

/** Optional base type for selection source: the event emitter interface is not required. */
export type SelectionSourceEventBase = EventSource<SelectionSourceEvents>;

/**
 * The user has selected an extent.
 */
export interface ExtentSelection {
    type: "extent";
    extent: Extent;
}

/**
 * The selection made by the user.
 *
 * This is currently always `type: "extent"`, but additional selection kinds
 * may be added in the future.
 *
 * Selection sources should check the `type` and throw an error for unsupported
 * selection kinds in order to remain forwards compatible.
 */
export type SelectionKind = ExtentSelection;

/**
 * An object that allows spatial selection.
 *
 * Developers can create classes that implement this interface for different selection sources.
 *
 * The implementation of `SelectionSourceEventBase` is optional: it is only necessary if the status changes
 * during the lifetime of the selection source.
 * To implement events, you can write `class MySelectionSource extends EventEmitter<SelectionSourceEvents>`.
 *
 */
export interface SelectionSource extends Partial<SelectionSourceEventBase> {
    /**
     * The label of this source.
     *
     * This will be displayed by the user interface during selection source selection.
     */
    readonly label: string;

    /**
     * The optional status of this source. If there is no status defined, it is assumed that the
     * source is always available.
     *
     * This will be displayed by the user interface.
     */
    readonly status?: SelectionSourceStatus;

    /**
     * Performs a selection and returns a list of selection results.
     *
     * Implementations should return the results ordered by priority (best match first), if possible.
     *
     * @param selectionKind The geometry with which to perform the spatial selection. Currently only
     * an extent is supported.
     * @param options see interface documentation {@link SelectionOptions}
     */
    select(selectionKind: SelectionKind, options: SelectionOptions): Promise<SelectionResult[]>;
}

export interface VectorLayerSelectionSourceOptions {
    vectorLayer: VectorLayer<Feature>;
    label: string;
}

export interface VectorLayerSelectionSource extends Required<SelectionSource>, Resource {}

/**
 * A factory that creates {@link VectorLayerSelectionSource | selection sources} to be used on an
 * OpenLayers VectorLayer with an OpenLayers VectorSource (e.g. layer of the map).
 *
 * Use the interface name `"selection.VectorSelectionSourceFactory"` to obtain an instance of this factory.
 */
export interface VectorLayerSelectionSourceFactory
    extends DeclaredService<"selection.VectorSelectionSourceFactory"> {
    /**
     * Returns a new {@link VectorLayerSelectionSource} that operates on the given OpenLayers VectorLayer.
     */
    createSelectionSource(options: VectorLayerSelectionSourceOptions): VectorLayerSelectionSource;
}
