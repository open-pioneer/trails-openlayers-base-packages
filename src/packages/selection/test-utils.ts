// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Reactive, reactive } from "@conterra/reactivity-core";
import {
    SelectionSource,
    SelectionResult,
    SelectionOptions,
    SelectionSourceStatus,
    SelectionKind
} from "./api";
import { Point } from "ol/geom";

export const fakeSelectedPointFeatures = [
    new Point([407354, 5754673]), // con terra (Bottom Right)
    new Point([404740, 5757893]) // Schloss (Top Left)
];

let count = 0;

export class FakePointSelectionSource implements SelectionSource {
    readonly label = `Fake Selection Source #${++count}`;
    #timeout: number;
    #status: Reactive<SelectionSourceStatus>;
    #pointFeatures: Point[];

    constructor(
        timeout?: number,
        status?: SelectionSourceStatus,
        pointFeatures: Point[] = fakeSelectedPointFeatures
    ) {
        this.#timeout = timeout || 0;
        this.#status = reactive(status || "unavailable");
        this.#pointFeatures = pointFeatures;
    }

    get status(): SelectionSourceStatus {
        return this.#status.value;
    }

    set status(value: SelectionSourceStatus) {
        this.#status.value = value;
    }

    async select(
        selectionKind: SelectionKind,
        options: SelectionOptions
    ): Promise<SelectionResult[]> {
        if (selectionKind.type !== "extent") {
            throw new Error(`Unsupported selection kind: ${selectionKind.type}`);
        }

        if (this.#status.value !== "available") return [];

        await new Promise((resolve) => setTimeout(resolve, this.#timeout));

        const allPoints = this.#pointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            if (!point.intersectsExtent(selectionKind.extent)) {
                return undefined;
            }
            return result;
        });

        const selectedPoints = allPoints.filter((s): s is SelectionResult => s != null);
        const limitedResults =
            selectedPoints.length > options.maxResults
                ? selectedPoints.slice(0, options.maxResults)
                : selectedPoints;
        return limitedResults;
    }
}

/**
 * For testing purposes only
 */
export class NoStatusSelectionSource implements SelectionSource {
    readonly label: string = "Testlabel";

    async select(_: SelectionKind, __: SelectionOptions): Promise<SelectionResult[]> {
        const allPoints = fakeSelectedPointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            return result;
        });
        return allPoints;
    }
}
