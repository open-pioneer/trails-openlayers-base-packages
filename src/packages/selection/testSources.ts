// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    SelectionSource,
    SelectionResult,
    SelectionOptions,
    SelectionSourceStatus,
    SelectionSourceEvents,
    SelectionKind
} from "./api";
import { Point } from "ol/geom";
import { EventEmitter } from "@open-pioneer/core";
import VectorSource from "ol/source/Vector";
import VectorLayer from "ol/layer/Vector";
import { EventsKey } from "ol/events";
import { unByKey } from "ol/Observable";

export const fakeSelectedPointFeatures = [
    new Point([407354, 5754673]), // con terra (Bottom Right)
    new Point([404740, 5757893]) // Schloss (Top Left)
];

let count = 0;

export class FakePointSelectionSource
    extends EventEmitter<SelectionSourceEvents>
    implements SelectionSource
{
    readonly label = `Fake Selection Source #${++count}`;
    #timeout: number;
    #status: SelectionSourceStatus;
    #pointFeatures: Point[];

    constructor(
        timeout?: number,
        status?: SelectionSourceStatus,
        pointFeatures: Point[] = fakeSelectedPointFeatures
    ) {
        super();
        this.#timeout = timeout || 0;
        this.#status = status || "unavailable";
        this.#pointFeatures = pointFeatures;
    }

    get status(): SelectionSourceStatus {
        return this.#status;
    }

    set status(value: SelectionSourceStatus) {
        if (value !== this.#status) {
            this.#status = value;
            this.emit("changed:status");
        }
    }

    async select(selection: SelectionKind, options: SelectionOptions): Promise<SelectionResult[]> {
        if (selection.type !== "extent") {
            throw new Error(`Unsupported selection kind: ${selection.type}`);
        }

        if (this.#status !== "available") return [];

        await new Promise((resolve) => setTimeout(resolve, this.#timeout));

        const allPoints = this.#pointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            if (!point.intersectsExtent(selection.extent)) {
                return undefined;
            }
            return result;
        });

        const selectedPoints = allPoints.filter((s): s is SelectionResult => s != null);
        const limitedResults =
            selectedPoints.length > options.maxResults
                ? selectedPoints.slice(0, options.maxResults)
                : selectedPoints;
        return limitedResults;
    }
}

export class VectorLayerSelectionSource
    extends EventEmitter<SelectionSourceEvents>
    implements SelectionSource
{
    readonly label: string;
    #status: Exclude<SelectionSourceStatus, string> = { kind: "available" };
    #vectorLayer: VectorLayer<VectorSource>;
    #eventHandler: EventsKey;
    #layerNotVisibleReason: string;

    constructor(
        vectorLayer: VectorLayer<VectorSource>,
        label: string,
        layerNotVisibleReason: string
    ) {
        super();
        this.label = label;
        this.#vectorLayer = vectorLayer;
        this.#layerNotVisibleReason = layerNotVisibleReason;
        this.#updateStatus();
        this.#eventHandler = this.#vectorLayer.on("change:visible", () => {
            this.#updateStatus();
        });
    }

    destroy() {
        unByKey(this.#eventHandler);
    }

    get status(): SelectionSourceStatus {
        return this.#status;
    }

    async select(selection: SelectionKind, options: SelectionOptions): Promise<SelectionResult[]> {
        if (selection.type !== "extent") {
            throw new Error(`Unsupported selection kind: ${selection.type}`);
        }

        if (this.#status.kind !== "available" || this.#vectorLayer.getSource() === null) return [];

        const allResults: SelectionResult[] = [];
        this.#vectorLayer
            .getSource()!
            .forEachFeatureIntersectingExtent(selection.extent, (feature) => {
                if (!feature.getGeometry()) return;
                const result: SelectionResult = {
                    id: feature.getId()?.toString() || feature.getGeometry.toString(),
                    geometry: feature.getGeometry()!
                };
                allResults.push(result);
            });
        const selectedFeatures = allResults.filter((s): s is SelectionResult => s != null);
        const limitedFeatures =
            selectedFeatures.length > options.maxResults
                ? selectedFeatures.slice(0, options.maxResults)
                : selectedFeatures;
        return limitedFeatures;
    }

    #updateStatus() {
        const layerIsVisible = this.#vectorLayer.getVisible();
        const newStatus: SelectionSourceStatus = layerIsVisible
            ? { kind: "available" }
            : { kind: "unavailable", reason: this.#layerNotVisibleReason };
        if (newStatus.kind !== this.#status.kind) {
            this.#status = newStatus;
            this.emit("changed:status");
        }
    }
}

export class NoStatusSelectionSource
    extends EventEmitter<SelectionSourceEvents>
    implements SelectionSource
{
    readonly label: string = "Testlabel";

    async select(_: SelectionKind, __: SelectionOptions): Promise<SelectionResult[]> {
        const allPoints = fakeSelectedPointFeatures.map((point, index) => {
            const result: SelectionResult = {
                id: index,
                geometry: point
            };
            return result;
        });
        return allPoints;
    }
}
