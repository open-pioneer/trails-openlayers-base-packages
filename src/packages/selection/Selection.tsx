// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    Box,
    Flex,
    FormControl,
    FormLabel,
    Icon,
    Tooltip,
    VStack,
    chakra,
    useToken
} from "@open-pioneer/chakra-integration";
import { MapModel, useMapModel } from "@open-pioneer/map";
import { NotificationService } from "@open-pioneer/notifier";
import { CommonComponentProps, useCommonComponentProps, useEvent } from "@open-pioneer/react-utils";
import { PackageIntl } from "@open-pioneer/runtime";
import {
    OptionProps,
    Select,
    Props as SelectProps,
    SingleValueProps,
    chakraComponents,
    type SingleValue,
    ChakraStylesConfig,
    GroupBase
} from "chakra-react-select";
import { Geometry } from "ol/geom";
import { useIntl, useService } from "open-pioneer:react-hooks";
import { FC, useCallback, useEffect, useMemo, useState } from "react";
import { FiAlertTriangle } from "react-icons/fi";
import { DragController } from "./DragController";
import { SelectionController } from "./SelectionController";
import { SelectionResult, SelectionSource, SelectionSourceStatusObject } from "./api";

/**
 * Properties supported by the {@link Selection} component.
 */
export interface SelectionProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;

    /**
     * Array of selection sources available for spatial selection.
     */
    sources: SelectionSource[];

    /**
     * This handler is called whenever the user has successfully selected
     * some items.
     */
    onSelectionComplete?(event: SelectionCompleteEvent): void;

    /**
     * This handler is called whenever the user has changed the selected source
     */
    onSelectionSourceChanged?(event: SelectionSourceChangedEvent): void;
}

export interface SelectionCompleteEvent {
    /** The source that returned the {@link results}. */
    source: SelectionSource;

    /** Results selected by the user. */
    results: SelectionResult[];
}

export interface SelectionSourceChangedEvent {
    /** The new selected source */
    source: SelectionSource | undefined;
}

/**
 * Properties for single select options.
 */
interface SelectionOption {
    /**
     * The label of the selection source option.
     */
    label: string;

    /**
     * The value (SelectionSource) of the selection source option.
     */
    value: SelectionSource | undefined;
}

/**
 * Properties for single selection method options.
 */
interface MethodOption {
    /**
     * The label of the select method option.
     */
    label: string;

    /**
     * The value of the select method option.
     */
    value: string;
}

/**
 * Supported selection methods
 */
export enum SelectionMethods {
    extent = "EXTENT",
    polygon = "POLYGON",
    free = "FREEPOLYGON",
    circle = "CIRCLE"
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const COMMON_SELECT_PROPS: SelectProps<any, any, any> = {
    classNamePrefix: "react-select",
    menuPosition: "fixed",
    isSearchable: false,
    isClearable: false
};

/**
 * A component that allows the user to perform a spatial selection on a given set of {@link SelectionSource}.
 */
export const Selection: FC<SelectionProps> = (props) => {
    const intl = useIntl();
    const { mapId, sources, onSelectionComplete, onSelectionSourceChanged } = props;
    const { containerProps } = useCommonComponentProps("selection", props);
    const [currentSource, setCurrentSource] = useState<SelectionSource | undefined>(() =>
        sources.find((s) => (s.status ?? "available") === "available")
    );
    const mapState = useMapModel(mapId);
    const { onExtentSelected } = useSelectionController(
        mapState.map,
        sources,
        currentSource,
        onSelectionComplete
    );
    const chakraStyles = useChakraStyles();

    /**
     * Method to build Option-Array from the supported selection methods for the selection-method react-select
     * If there is no configuration => Default selection method: EXTENT
     */
    const buildMethodOptions = useCallback(
        (methods: string[] | undefined) => {
            const objects: MethodOption[] = [];
            if (!methods) methods = [SelectionMethods.extent];
            methods.forEach((item) => {
                if (Object.values(SelectionMethods as unknown as string[]).includes(item))
                    objects.push({ label: intl.formatMessage({ id: item }), value: item });
            });
            if (objects.length === 0) throw new Error("methods does not contain valid values");
            return objects;
        },
        [intl]
    );

    const methodOptions: MethodOption[] = buildMethodOptions(undefined);
    const [selectedMethod, setSelectedMethod] = useState(methodOptions[0] as MethodOption);

    /**
     * Method to change used selectmethod
     */
    const onMethodeOptionChance = useEvent((newValue: MethodOption) => {
        setSelectedMethod(newValue);
    });

    const [dragControllerActive, setDragControllerActive] = useState<boolean>(true);
    useDragSelection(mapState.map, selectedMethod, intl, onExtentSelected, dragControllerActive);

    /**
     * Method to build Option-Array from sources for the selection-source react-select
     */
    const sourceOptions = useMemo(
        () =>
            sources.map<SelectionOption>((source) => {
                return { label: source.label, value: source };
            }),
        [sources]
    );
    const currentSourceOption = useMemo(
        () => sourceOptions.find((option) => option.value === currentSource),
        [sourceOptions, currentSource]
    );

    /**
     * Method to change used source
     */
    const onSourceOptionChanged = useEvent((newValue: SingleValue<SelectionOption>) => {
        setCurrentSource(newValue?.value);
        onSelectionSourceChanged && onSelectionSourceChanged({ source: newValue?.value });
    });

    useEffect(() => {
        if (!currentSource) {
            setDragControllerActive(false);
            return;
        }

        const sourceNotAvailableReason = intl.formatMessage({ id: "sourceNotAvailable" });
        const isCurrentSourceAvailable = () => {
            return (
                currentSource &&
                getSourceStatus(currentSource, sourceNotAvailableReason).kind === "available"
            );
        };

        setDragControllerActive(isCurrentSourceAvailable());
        // Why can this be undefined after test above?!
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const handle = currentSource.on("changed:status", () => {
            setDragControllerActive(isCurrentSourceAvailable());
        });
        return () => handle.destroy();
    }, [currentSource, setDragControllerActive, intl]);

    return (
        <VStack {...containerProps} spacing={2}>
            {methodOptions.length > 1 && (
                <FormControl>
                    <FormLabel>{intl.formatMessage({ id: "selectMethod" })}</FormLabel>
                    <Select
                        className="selection-method react-select"
                        {...COMMON_SELECT_PROPS}
                        options={methodOptions}
                        onChange={onMethodeOptionChance}
                        value={selectedMethod}
                        chakraStyles={chakraStyles}
                    />
                </FormControl>
            )}
            <FormControl>
                <FormLabel>{intl.formatMessage({ id: "selectSource" })}</FormLabel>
                <Select<SelectionOption>
                    className="selection-source react-select"
                    {...COMMON_SELECT_PROPS}
                    options={sourceOptions}
                    placeholder={intl.formatMessage({ id: "selectionPlaceholder" })}
                    value={currentSourceOption}
                    onChange={onSourceOptionChanged}
                    components={{
                        Option: SourceSelectOption,
                        SingleValue: SourceSelectValue
                    }}
                    isOptionDisabled={(option) =>
                        option.value === undefined || option.value.status === "unavailable"
                    }
                    // optionLabel is used by screenreaders
                    getOptionLabel={(option) =>
                        option.label +
                        (option.value === undefined || option.value.status === "unavailable"
                            ? " " + intl.formatMessage({ id: "sourceNotAvailable" })
                            : "")
                    }
                    chakraStyles={chakraStyles}
                />
            </FormControl>
        </VStack>
    );
};

function SourceSelectOption(props: OptionProps<SelectionOption>): JSX.Element {
    const { value } = props.data;
    const { isAvailable, content } = useSourceItem(value, false);

    return (
        <chakraComponents.Option
            {...props}
            isDisabled={!isAvailable}
            className="selection-source-option"
        >
            {content}
        </chakraComponents.Option>
    );
}

function SourceSelectValue(props: SingleValueProps<SelectionOption>): JSX.Element {
    const { value } = props.data;
    const { isAvailable, content } = useSourceItem(value, true);
    const clazz = isAvailable
        ? "selection-source-value"
        : "selection-source-value selection-source-value--disabled";

    return (
        <chakraComponents.SingleValue {...props} isDisabled={!isAvailable} className={clazz}>
            {content}
        </chakraComponents.SingleValue>
    );
}

/**
 * Hook to manage source option in selection-source react-select
 */
function useSourceItem(source: SelectionSource | undefined, isSelected: boolean) {
    const label: string | undefined = source?.label;
    const status = useSourceStatus(source);

    return {
        isAvailable: status.kind === "available",
        content: (
            <Flex direction="row" alignItems="center" grow={1}>
                {!isSelected && <Flex grow={1}>{label}</Flex>}
                {status.kind === "unavailable" && (
                    <Box ml={2}>
                        <Tooltip label={status.reason} placement="right" openDelay={500}>
                            <chakra.span>
                                <Icon
                                    as={FiAlertTriangle}
                                    color="red"
                                    className="warning-icon"
                                    aria-label={status.reason}
                                />
                            </chakra.span>
                        </Tooltip>
                    </Box>
                )}
                {isSelected && label}
            </Flex>
        )
    };
}

/**
 * Hook to manage selection sources
 */
function useSelectionController(
    mapModel: MapModel | undefined,
    sources: SelectionSource[],
    currentSource: SelectionSource | undefined,
    onSelectionComplete: ((event: SelectionCompleteEvent) => void) | undefined
) {
    const notifier = useService<NotificationService>("notifier.NotificationService");
    const intl = useIntl();
    const [controller, setController] = useState<SelectionController | undefined>(undefined);
    useEffect(() => {
        if (!mapModel) {
            return;
        }
        const controller = new SelectionController({
            mapModel,
            onError() {
                notifier.notify({
                    level: "error",
                    message: intl.formatMessage({ id: "selectionFailed" })
                });
            }
        });
        setController(controller);
        return () => {
            controller.destroy();
        };
    }, [mapModel, notifier, sources, intl]);

    const onExtentSelected = useEvent(async (geometry: Geometry) => {
        if (!controller || !currentSource) {
            return;
        }

        const selectionResult = await controller.select(currentSource, geometry.getExtent());
        if (!selectionResult) {
            return;
        }

        onSelectionComplete?.(selectionResult);
    });
    return {
        controller,
        onExtentSelected
    };
}

type SimpleStatus =
    | {
          kind: "available";
      }
    | {
          kind: "unavailable";
          reason: string;
      };

function getSourceStatus(source: SelectionSource, sourceNotAvailableReason: string): SimpleStatus {
    const rawCurrent = source.status ?? "available";
    const current: SelectionSourceStatusObject =
        typeof rawCurrent === "string" ? { kind: rawCurrent } : rawCurrent;
    if (current.kind === "available") {
        return current;
    }

    return {
        kind: "unavailable",
        reason: current.reason ?? sourceNotAvailableReason
    };
}

/**
 * Hook to manage source status
 */
function useSourceStatus(source: SelectionSource | undefined): SimpleStatus {
    const intl = useIntl();
    const [status, setStatus] = useState<SimpleStatus>(() => ({ kind: "available" }));
    useEffect(() => {
        if (!source) {
            setStatus({ kind: "available" });
            return;
        }
        const sourceNotAvailableReason = intl.formatMessage({ id: "sourceNotAvailable" });
        setStatus(getSourceStatus(source, sourceNotAvailableReason));
        const resource = source.on?.("changed:status", () => {
            setStatus(getSourceStatus(source, sourceNotAvailableReason));
        });
        return () => resource?.destroy();
    }, [source, intl]);
    return status;
}

/**
 * Hook to manage map controls and tooltip
 */
function useDragSelection(
    map: MapModel | undefined,
    selectMethode: MethodOption,
    intl: PackageIntl,
    onExtentSelected: (geometry: Geometry) => void,
    isActive: boolean
) {
    useEffect(() => {
        if (!map) {
            return;
        }

        const dragController = new DragController(
            map.olMap,
            selectMethode.value,
            intl.formatMessage({ id: "tooltip" }),
            intl.formatMessage({ id: "disabledTooltip" }),
            onExtentSelected
        );
        dragController.setActive(isActive);

        return () => {
            dragController?.destroy();
        };
    }, [map, selectMethode, intl, onExtentSelected, isActive]);
}

/**
 * Customizes components styles within the select component.
 */
function useChakraStyles() {
    const [dropDownBackground, borderColor] = useToken(
        "colors",
        ["background_body", "border"],
        ["#ffffff", "#ffffff"]
    );
    return useMemo(() => {
        const chakraStyles: ChakraStylesConfig<
            SelectionOption,
            false,
            GroupBase<SelectionOption>
        > = {
            control: (styles) => ({ ...styles, cursor: "pointer" }),
            indicatorSeparator: (styles) => ({
                ...styles,
                borderColor: borderColor
            }),
            dropdownIndicator: (provided) => ({
                ...provided,
                backgroundColor: dropDownBackground
            })
        };
        return chakraStyles;
    }, [dropDownBackground, borderColor]);
}
