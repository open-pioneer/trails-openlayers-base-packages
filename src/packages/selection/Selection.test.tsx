// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0

import { afterEach, expect, it, vi } from "vitest";
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import { Selection, SelectionCompleteEvent, SelectionSourceChangedEvent } from "./Selection";
import { FakePointSelectionSource, NoStatusSelectionSource } from "./testSources";
import { VectorLayerSelectionSource } from "./testSources";
import { NotificationService } from "@open-pioneer/notifier";
import { Point } from "ol/geom";
import { SelectionSource } from "./api";
import VectorSource from "ol/source/Vector";
import GeoJSON from "ol/format/GeoJSON";
import VectorLayer from "ol/layer/Vector";
import { SimpleLayer } from "@open-pioneer/map";

afterEach(() => {
    vi.restoreAllMocks();
});

it("should successfully create a selection component", async () => {
    await createSelection();
    const { selectionDiv } = await waitForSelection();
    expect(selectionDiv).toMatchSnapshot();
});

it("Sources read from component should be the same as the provided sources", async () => {
    const ALL_EXPECTED_FEATURES = [new Point([407354, 5754673]), new Point([404740, 5757893])];
    const POINT_SOURCE = new FakePointSelectionSource(0, "available", ALL_EXPECTED_FEATURES);

    await createSelection([POINT_SOURCE]);
    const { selectElement } = await waitForSelection();
    openOptions(selectElement);

    const options = getOptions(selectElement);
    const option = options[0];

    expect(options.length).equals(1);
    expect(option?.textContent).toBe(POINT_SOURCE.label);
});

it("Should disable option and show warning icon for unavailable sources", async () => {
    const ALL_EXPECTED_FEATURES = [new Point([407354, 5754673]), new Point([404740, 5757893])];
    const POINT_SOURCE = new FakePointSelectionSource(0, "unavailable", ALL_EXPECTED_FEATURES);

    await createSelection([POINT_SOURCE]);
    const { selectionDiv, selectElement } = await waitForSelection();
    openOptions(selectElement);

    const options = getOptions(selectElement);
    const option = options[0];
    const warningIcon = selectionDiv.querySelector(".warning-icon");

    expect(warningIcon).not.toBeNull();
    expect(warningIcon).toBeDefined();
    expect(option?.classList.contains("react-select__option--is-disabled")).toBeTruthy();
});

it("Should disable or enable selection option when changing the status of a source layer", async () => {
    const layer = new SimpleLayer({
        id: "ogc_kitas",
        title: "Kindertagesst√§tten",
        visible: false,
        olLayer: createKitasLayer()
    });

    layer.olLayer.setVisible(false);

    const layerSelectionSource = new VectorLayerSelectionSource(
        layer.olLayer as VectorLayer<VectorSource>,
        layer.title,
        "Layer not visible"
    );
    await createSelection([layerSelectionSource]);
    const { selectElement } = await waitForSelection();
    openOptions(selectElement);

    const options = getOptions(selectElement);
    const option = options[0];
    expect(option?.classList.contains("react-select__option--is-disabled")).toBeTruthy();

    act(() => {
        layer.olLayer.setVisible(true);
    });
    openOptions(selectElement);
    expect(option?.classList.contains("react-select__option--is-disabled")).toBeFalsy();
});

it("expect selection source with no defined status is still available", async () => {
    const noStatusSelectionSource = new NoStatusSelectionSource();
    await createSelection([noStatusSelectionSource]);
    const { selectElement } = await waitForSelection();
    openOptions(selectElement);

    const options = getOptions(selectElement);
    const option = options[0];
    expect(option?.classList.contains("react-select__option--is-disabled")).toBeFalsy();
});

async function createSelection(selectionSources?: SelectionSource[] | undefined) {
    const { mapId, registry } = await setupMap();

    const notifier: Partial<NotificationService> = {
        notify() {
            throw new Error("not implemented");
        }
    };

    const injectedServices = createServiceOptions({
        registry
    });
    injectedServices["notifier.NotificationService"] = notifier;
    const sources = selectionSources || [new FakePointSelectionSource()];
    render(
        <PackageContextProvider services={injectedServices}>
            <Selection
                data-testid="selection"
                mapId={mapId}
                sources={sources}
                onSelectionComplete={onSelectionComplete}
                onSelectionSourceChanged={onSelectionSourceChanged}
            ></Selection>
        </PackageContextProvider>
    );
}

async function waitForSelection() {
    const { selectionDiv, selectElement } = await waitFor(async () => {
        const selectionDiv = await screen.findByTestId<HTMLDivElement>("selection");
        if (!selectionDiv) {
            throw new Error("Selection not rendered");
        }

        const selectElement: HTMLSelectElement | null =
            selectionDiv.querySelector(".selection-source");
        if (!selectElement) {
            throw new Error("Select element not rendered");
        }

        return { selectionDiv, selectElement };
    });

    return { selectionDiv, selectElement };
}

function openOptions(selectElement: HTMLSelectElement) {
    act(() => {
        fireEvent.keyDown(selectElement, { key: "ArrowDown" });
    });
}

function getOptions(selectElement: HTMLSelectElement) {
    return Array.from(
        selectElement.getElementsByClassName("selection-source-option")
    ) as HTMLElement[];
}

function createKitasLayer() {
    const geojsonSource = new VectorSource({
        url: "https://ogc-api.nrw.de/inspire-us-kindergarten/v1/collections/governmentalservice/items?f=json&limit=10000",
        format: new GeoJSON(), //assign GeoJson parser
        attributions:
            '&copy; <a href="http://www.bkg.bund.de" target="_blank">Bundesamt f&uuml;r Kartographie und Geod&auml;sie</a> 2017, <a href="http://sg.geodatenzentrum.de/web_public/Datenquellen_TopPlus_Open.pdf" target="_blank">Datenquellen</a>'
    });

    return new VectorLayer({
        source: geojsonSource
    });
}

function onSelectionComplete(event: SelectionCompleteEvent) {
    const geometries = event.results.map((result) => result.geometry);
    return geometries;
}

function onSelectionSourceChanged(_: SelectionSourceChangedEvent) {
    return;
}
