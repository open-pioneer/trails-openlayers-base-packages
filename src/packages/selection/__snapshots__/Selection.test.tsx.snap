// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a selection component 1`] = `
<div
  class="chakra-stack selection css-ldoqu4"
  data-testid="selection"
  data-theme="light"
>
  <div
    class="chakra-form-control css-1kxonj9"
    data-theme="light"
    role="group"
  >
    <label
      class="chakra-form__label css-g6pte"
      data-theme="light"
      for="field-:r0:"
      id="field-:r0:-label"
    >
      selectSource
    </label>
    <div
      class="selection-source react-select css-79elbk"
      data-theme="light"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-2-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        class="react-select__control css-156sfem"
        data-theme="light"
      >
        <div
          class="react-select__value-container css-j93siq"
          data-theme="light"
        >
          <div
            class="react-select__placeholder css-hji3d9"
            data-theme="light"
            id="react-select-2-placeholder"
          >
            selectionPlaceholder
          </div>
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-describedby="react-select-2-placeholder"
            aria-expanded="false"
            aria-haspopup="true"
            aria-readonly="true"
            aria-required="false"
            class="css-mohuvp-dummyInput-DummyInput"
            id="field-:r0:"
            inputmode="none"
            role="combobox"
            tabindex="0"
            value=""
          />
        </div>
        <div
          class="react-select__indicators css-hfbj6y"
          data-theme="light"
        >
          <hr
            aria-orientation="vertical"
            class="chakra-divider react-select__indicator-separator css-1611kif"
            data-theme="light"
          />
          <div
            aria-hidden="true"
            class="react-select__indicator react-select__dropdown-indicator css-1lza9k0"
            data-theme="light"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              data-theme="light"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
